source/system/TextDatasource.js

```diff
@@ -2,18 +2,95 @@
+ * @type {Function}
+ * @private
+ */
+let __appointedEncToHistoryCB = convertEncryptedContentToHistory,
+    /**
+     * Current appointed callback for encrypting archive history
+     * @type {Function}
+     * @private
+     */
+    __appointedHistoryToEncCB = convertHistoryToEncryptedContent;
+
+/**
+ * Convert encrypted text to an array of commands (history)
+ * @param {String} encText The encrypted archive content
+ * @param {Credentials} credentials A credentials instance that has a password, keyfile
+ *  or both
+ * @returns {Promise.<Array>} A promise that resolves with an array of commands
+ */
+function convertEncryptedContentToHistory(encText, credentials) {
+    const { password, keyfile } = processCredentials(credentials);
+    return Promise.resolve(encText)
+        .then(function __stripSignature(data) {
+            if (!signing.hasValidSignature(data)) {
+                throw new Error("No valid signature in archive");
+            }
+            return signing.stripSignature(data);
+        })
+        .then(function __decryptUsingKeyFile(encryptedData) {
+            // optionally decrypt using a key file
+            return keyfile ?
+                iocane.decryptWithKeyFile(encryptedData, keyfile) :
+                encryptedData;
+        })
+        .then(function __decryptUsingPassword(encryptedData) {
+            // optionally decrypt using a password
+            return password ?
+                iocane.decryptWithPassword(encryptedData, password) :
+                encryptedData;
+        })
+        .then(function __marshallHistoryToArray(decrypted) {
+            if (decrypted && decrypted.length > 0) {
+                var decompressed = encoding.decompress(decrypted);
+                if (decompressed) {
+                    return historyTools.historyStringToArray(decompressed);
+                }
+            }
+            throw new Error("Decryption failed");
+        });
+}
+
+/**
+ * Convert an array of commands (history) to an encrypted string
+ * @param {Array.<String>} historyArr An array of commands
+ * @param {Credentials} credentials A credentials instance that has a password, keyfile
+ *  or both
+ * @returns {String} Encrypted archive contents
+ */
+function convertHistoryToEncryptedContent(historyArr, credentials) {
+    const { password, keyfile } = processCredentials(credentials);
+    const history = historyTools.historyArrayToString(historyArr);
+    const compressed = encoding.compress(history);
+    return Promise
+        .resolve(compressed)
+        .then(function __encryptUsingPassword(encryptedData) {
+            return password ?
+                iocane.encryptWithPassword(encryptedData, password) :
+                encryptedData;
+        })
+        .then(function __encryptUsingKeyFile(encryptedData) {
+            return keyfile ?
+                iocane.encryptWithKeyFile(encryptedData, keyfile) :
+                encryptedData;
+        })
+        .then(signing.sign);
```

Comment:
Yeah.. the crypto around pass+keyfile is interesting. It's basically password first, keyfile second (double encryption). If neither are provided it should [fail here](https://github.com/buttercup/buttercup-core/pull/157/files/f17c86ee8ea0234d1c4761317b03eb98ca90c141#diff-28be79e663ac9e4b2428d31532ea916cR109).